File: apps/api/src/f1hub/api/v1/predictions.py
"""
Prediction API Endpoints

Endpoints for making ML model predictions.
"""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from minio import Minio
from sqlalchemy.orm import Session

from f1hub.core.dependencies import get_db, get_minio_client
from f1hub.schemas.predictions import (
    LapTimePredictionResponse,
    OvertakePredictionResponse,
    PredictionRequest,
    TyreDegradationPredictionResponse,
)
from f1hub.services.ml_service import MLService

router = APIRouter(prefix="/predictions")


@router.get("/tyre-degradation/{stint_id}", response_model=TyreDegradationPredictionResponse)
async def predict_tyre_degradation(
    stint_id: UUID,
    db: Session = Depends(get_db),
    minio: Minio = Depends(get_minio_client),
) -> TyreDegradationPredictionResponse:
    """Predict tyre degradation for a stint.

    Args:
        stint_id: Stint UUID
        db: Database session
        minio: MinIO client

    Returns:
        Prediction with deg_per_lap value

    Raises:
        HTTPException: If stint not found or model not available
    """
    try:
        ml_service = MLService(db, minio)
        result = ml_service.predict_tyre_degradation(stint_id)

        return TyreDegradationPredictionResponse(**result)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Prediction failed: {str(e)}")


@router.get("/lap-time", response_model=LapTimePredictionResponse)
async def predict_lap_time(
    tyre_age: int,
    compound: str,
    track_status: str,
    position: int,
    driver_id: str,
    db: Session = Depends(get_db),
    minio: Minio = Depends(get_minio_client),
) -> LapTimePredictionResponse:
    """Predict lap time for given conditions.

    Args:
        tyre_age: Laps on current tyre
        compound: Tyre compound (SOFT, MEDIUM, HARD)
        track_status: Track status (GREEN, YELLOW, SC, VSC, RED)
        position: Current position
        driver_id: Driver identifier
        db: Database session
        minio: MinIO client

    Returns:
        Prediction with lap time in seconds

    Raises:
        HTTPException: If model not available
    """
    try:
        ml_service = MLService(db, minio)
        result = ml_service.predict_lap_time(
            tyre_age, compound, track_status, position, driver_id
        )

        return LapTimePredictionResponse(**result)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Prediction failed: {str(e)}")


@router.get("/overtake", response_model=OvertakePredictionResponse)
async def predict_overtake(
    gap_seconds: float,
    closing_rate: float,
    tyre_advantage: int,
    drs_available: bool,
    lap_number: int,
    db: Session = Depends(get_db),
    minio: Minio = Depends(get_minio_client),
) -> OvertakePredictionResponse:
    """Predict overtake probability.

    Args:
        gap_seconds: Time gap between cars (seconds)
        closing_rate: Gap change per lap (s/lap, negative = closing)
        tyre_advantage: Compound advantage (-2 to +2)
        drs_available: Whether DRS is available
        lap_number: Current lap number
        db: Database session
        minio: MinIO client

    Returns:
        Prediction with overtake probability (0-1)

    Raises:
        HTTPException: If model not available
    """
    try:
        ml_service = MLService(db, minio)
        result = ml_service.predict_overtake(
            gap_seconds, closing_rate, tyre_advantage, drs_available, lap_number
        )

        return OvertakePredictionResponse(**result)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Prediction failed: {str(e)}")

File: apps/web/src/lib/api/predictions.ts
/**
 * Predictions API Client
 *
 * Functions for interacting with ML prediction endpoints.
 */

import { apiGet } from "./client";
import type {
  Model,
  TyreDegradationPrediction,
  LapTimePrediction,
  OvertakePrediction,
  RaceResultPrediction,
} from "@/types/predictions";

/**
 * Fetch all registered models
 */
export async function fetchModels(
  modelName?: string,
  status?: string
): Promise<Model[]> {
  const params = new URLSearchParams();
  if (modelName) params.append("model_name", modelName);
  if (status) params.append("status", status);

  const response = await apiGet<{ models: Model[] }>(
    `/models${params.toString() ? `?${params}` : ""}`
  );
  return response.models;
}

/**
 * Fetch a specific model by ID
 */
export async function fetchModel(modelId: string): Promise<Model> {
  return apiGet<Model>(`/models/${modelId}`);
}

/**
 * Predict tyre degradation for a stint
 */
export async function predictTyreDegradation(
  stintId: string
): Promise<TyreDegradationPrediction> {
  return apiGet<TyreDegradationPrediction>(
    `/predictions/tyre-degradation/${stintId}`
  );
}

/**
 * Predict lap time for given conditions
 */
export async function predictLapTime(params: {
  tyre_age: number;
  compound: string;
  track_status: string;
  position: number;
  driver_id: string;
}): Promise<LapTimePrediction> {
  const queryParams = new URLSearchParams({
    tyre_age: params.tyre_age.toString(),
    compound: params.compound,
    track_status: params.track_status,
    position: params.position.toString(),
    driver_id: params.driver_id,
  });

  return apiGet<LapTimePrediction>(`/predictions/lap-time?${queryParams}`);
}

/**
 * Predict overtake probability
 */
export async function predictOvertake(params: {
  gap_seconds: number;
  closing_rate: number;
  tyre_advantage: number;
  drs_available: boolean;
  lap_number: number;
}): Promise<OvertakePrediction> {
  const queryParams = new URLSearchParams({
    gap_seconds: params.gap_seconds.toString(),
    closing_rate: params.closing_rate.toString(),
    tyre_advantage: params.tyre_advantage.toString(),
    drs_available: params.drs_available.toString(),
    lap_number: params.lap_number.toString(),
  });

  return apiGet<OvertakePrediction>(`/predictions/overtake?${queryParams}`);
}

/**
 * Predict race result finishing position
 */
export async function predictRaceResult(params: {
  grid_position: number;
  avg_lap_time: number;
  driver_id: string;
}): Promise<RaceResultPrediction> {
  const queryParams = new URLSearchParams({
    grid_position: params.grid_position.toString(),
    avg_lap_time: params.avg_lap_time.toString(),
    driver_id: params.driver_id,
  });

  return apiGet<RaceResultPrediction>(`/predictions/race-result?${queryParams}`);
}
